presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,283 @@
+                    .build();
+            PlanNode left = getJoinSource(
+                    idAllocator,
+                    leftSources,
+                    pushDownResult.getLeftPredicate(),
+                    joinSymbols.stream().filter(leftSymbols::contains).collect(toImmutableList()));
+            PlanNode right = getJoinSource(
+                    idAllocator,
+                    rightSources,
+                    pushDownResult.getRightPredicate(),
+                    joinSymbols.stream()
+                            .filter(symbol -> !leftSymbols.contains(symbol))
+                            .collect(toImmutableList()));
+
+            List<Expression> joinPredicates = extractConjuncts(pushDownResult.getJoinPredicate());
+            List<JoinNode.EquiJoinClause> joinConditions = joinPredicates.stream()
+                    .filter(JoinEnumerator::isJoinEqualityCondition)
+                    .map(predicate -> toEquiJoinClause((ComparisonExpression) predicate, leftSymbols))
+                    .collect(toImmutableList());
+            List<Expression> joinFilters = joinPredicates.stream()
+                    .filter(predicate -> !isJoinEqualityCondition(predicate))
+                    .collect(toImmutableList());
+
+            // sort output symbols so that the left input symbols are first
+            List<Symbol> inputSymbols = ImmutableList.<Symbol>builder()
+                    .addAll(left.getOutputSymbols())
+                    .addAll(right.getOutputSymbols())
+                    .build();
+            List<Symbol> outputSymbols = inputSymbols.stream()
+                    .filter(joinGraph.getOutputSymbols()::contains)
+                    .collect(toImmutableList());
+            return new JoinNode(
+                    idAllocator.getNextId(),
+                    INNER,
+                    left,
+                    right,
+                    flipJoinCriteriaIfNecessary(joinConditions, left),
+                    outputSymbols,
+                    joinFilters.isEmpty() ? Optional.empty() : Optional.of(ExpressionUtils.and(joinFilters)),
+                    Optional.empty(),
+                    Optional.empty(),
+                    Optional.empty());
+        }
+
+        private static List<JoinNode.EquiJoinClause> flipJoinCriteriaIfNecessary(List<JoinNode.EquiJoinClause> joinCriteria, PlanNode left)
+        {
+            return joinCriteria
+                    .stream()
+                    .map(criterion -> left.getOutputSymbols().contains(criterion.getLeft()) ? criterion : criterion.flip())
+                    .collect(toImmutableList());
+        }
+
+        private static PlanNode getJoinSource(PlanNodeIdAllocator idAllocator, List<PlanNode> nodes, Expression filter, List<Symbol> outputSymbols)
+        {
+            PlanNode planNode;
+            if (nodes.size() == 1) {
+                planNode = getOnlyElement(nodes);
+                if (!(BooleanLiteral.TRUE_LITERAL).equals(filter)) {
```

Comment:
You can use equality inference to derive filters that can be rewritten using `planNode` output symbols (similar as for join)